### ISSUE TITLE
setup topcoder-react-utils as a dependency of Connect App, and adopt Babel configs from there

### ISSUE THREAD

**Issue by vic-topcoder** [appirio-tech/connect-app#1608]
Let's figure out how to add `topcoder-react-utils` into Connect and start the unification process for our apps.

From @birdofpreyru #1442 :
> Tt would be ideal to already try to setup topcoder-react-utils as a dependency of Connect App, and adopt Babel configs from there in Connect App (it might demand some work, cause those configs use Babel plugins to support CSS modules, thus any CSS code written before without CSS modules in mind should be wrapped into :global blocks, to keep it working).

@vikasrohit @gondzo @RishiRajSahu @maxceem @ThomasKranitsas @mtwomey FYI

How different is the environment and what would it take to do the step?

As a refactoring effort - can we also substitute our button implementations with the button component used in community app? We would need to document the process for the steps, so any challenge work should include a detailed user documentation. Also - I like how the code on community app is very well documented, can we try to get to a similar level of documentation in Connect?

Thanks!

**Comment by birdofpreyru** [comment]
For the button, I am still working on spawning this out of Community App. The JSX `<Button>` component itself is already moved to `topcoder-react-utils`, but moving related TC UI Kit to a separate repo, and also properly packing it for re-use in other apps, it will take me some more time.

**Comment by vic-topcoder** [comment]
Sounds like a great progress!

**Comment by vic-topcoder** [comment]
I also was playing with some react ui component libraries to see if we can grab and reuse any components from there for our needs.

Have you had any experience with Element http://element.eleme.io/#/en-US/component/button (they have a react port https://github.com/eleme/element-react/tree/master/src) or Ant Design https://ant.design/components/timeline/

**Comment by birdofpreyru** [comment]
C'mon @vic-topcoder , we are Topcoder community, we can create any components from scratch ourselves :)

We do use some 3-rd party components though to save time :)

I see you link specifically to the `<Button>` in Element? I believe, it is somewhat technically different from the `<Button>` we use in Community App. Our `<Button>`, under-the-hood, is a hybrid of `<Link>` (which in turns can be a simple `<a>` or React Router's `<Link>`) and `<button>`. The way it, and its themes, are done ensures that once it is correctly laid out into the page and styled, if one day you come to me and say: *this button should redirect user to another page inside the app*; and the next day you say *let's replace the redirection by some logic*, and yet another day *let's redirect user elsewhere and do some API call before the transition*; our `<Button>` allows to make such changes very easy, and without any care about its styling.

**Comment by vic-topcoder** [comment]
I am sure we can build everything :) Just want to look at what other people are doing and get the knowledge from libraries transferred to our library. The goal is not to reinvent the wheel and if there's a good component candidate we'd get the code. I hate code bloat and building a library of all different generic components is not great.

The button in Element has interesting properties we haven't designed/used - loading status of button, icons in button, segmented button.

I like the versatility of the button we have now, don't get me wrong :) Not suggesting we replace it at all!

Third-party components that we heavily use:
- React-select
... not sure if we have others

**Comment by birdofpreyru** [comment]
Without looking into the code, I recall that we also use `react-stickynode`, something for tooltips, `react-waypoint`, something for date/time selection; I guess something else.

**Comment by vic-topcoder** [comment]
@maxceem is going to add a utility - [react-popper](https://github.com/souporserious/react-popper) which wraps [popper.js](https://popper.js.org/) and now reimplementing Tooltip - for positioning of elements - tooltips, dropdowns - and this could potentially replace `react-stickynode` as the

**Comment by maxceem** [comment]
> How different is the environment and what would it take to do the step?

Regarding `topcoder-react-util` environments are almost same.
- The biggest difference is `CSS Modules` which `connect-app` doesn't use, but `community-app` does. So as birdofpreyru  said we will have to add `:global` to all existent CSS files in `connect-app` in order to works with `CSS Modules`
- Second thing is that `connect-app` requires `coffeescript` (which is required by its dependencies). And also there are other cases when custom modifications have to be applied to webpack config. For example, `connect-app` uses `handlebars` for rendering notifications. Handlebars has an issue which is fixed by adding 
    ```
    alias: {
      handlebars: 'handlebars/dist/handlebars.min.js'
    }
    ```
    Thus, probably `topcoder-react-util` has to support custom modifications which can be merged to default webpack configs. Previous package which played role of `topcoder-react-util` https://github.com/appirio-tech/webpack-config does support custom customization of webpack config if I'm not mistaken.

**Comment by birdofpreyru** [comment]
> Thus, probably `topcoder-react-util` has to support custom modifications

Sure, it does. For the Babel configs we are discussing now, they are done there as presets, and a Babel preset is just a JS object, holding lists of presets and plugins to use, their options and order. So, you can just import a preset provided by `topcoder-react-utils`, mutate it the way you need, and inject as your customized preset into Connect App's Babel config. However, this opportunity should not hide the fact that we want to eventually just reuse the very same config, so we need to gradually refactor Connect App for that (getting rid of Coffee Script dependencies, moving to CSS modules, etc.)

> Just remembered. As communit-app uses babel-plugin-css-modules-transform which is lightweight version of react-css-modules and doesn't cover all use cases. In theory there could be some other cases which will require additionaly adjustements. I don't remember what are restrictions of babel-plugin-css-modules-transform in comparison with react-css-modules.

Me neither, so I believe, in 95% cases it is nothing to worry about. The main difference is that `babel-plugin-css-modules-transform` makes all necessary class name substitutions at the compile time, rather that on the run time; but, I believe, when something cannot be resolved at the compile time, it still injects some helper functions that do the runtime resolution, so I don't know, whether there are any real restrictions, if any at all.

**Comment by birdofpreyru** [comment]
For the Webpack config, I am in progress of moving the standard part of it from Community App to `topcoder-react-utils`. It will sure demand some customization for each project using it, and it will be easy to do, thanks to `webpackMerge(..)`

**Comment by vic-topcoder** [comment]
Sounds great you guys! Let's keep pushing and plan the modifications done in Connect. @gondzo - do we need a new issue for all that's needed to include `topcoder-react-utils`?

**Comment by maxceem** [comment]
@gondzo currently to fix installed dependencies tree we use `npm-shrinkwrap.json` file. Can we remove it and use `package-lock.json` which is handled my npm@5 automatically? So we don't need to take care about it manually like it's described [in readme](https://github.com/appirio-tech/connect-app#updating-npm-shrinkwrapjson).

**Comment by gondzo** [comment]
yes, I think that's a good idea. we introduced shrinkwrap so we could have reproducible builds. package-lock serves the same purpose so we should be fine.

**Comment by vic-topcoder** [comment]
@mtwomey FYI

**Comment by maxceem** [comment]
Update.
`appirio-tech-react-components` package doesn't pre-build its modules and thus is being built with the same webpack config as `connect-app`.
So I'll have to also make `appirio-tech-react-components` package compatible with `topcoder-react-utils` webpack config.

**Comment by birdofpreyru** [comment]
I guess, it is easier to tune the config so that exclusively for `appirio-tech-react-components` it follows the old build rules? And then we just target to get rid of the `appirio-tech-react-components` as the dependency?

**Comment by maxceem** [comment]
> I guess, it is easier to tune the config so that exclusively for appirio-tech-react-components it follows the old build rules?

So far I'm trying to fix it rather than adjusting config. 

As I see in some situations adjusting small thing in config requires to repeat a big part of config and ruin the idea of reusing configs. 

Example:
- Some common styles in `appirio-tech-react-components` are imported by a "shortcut" : https://github.com/appirio-tech/react-components/blob/feature/connectv2/components/Dropdown/Dropdown.scss#L1
- This is achieved by the defining `includePaths` in webpack config: https://github.com/appirio-tech/connect-app/blob/dev/config/webpack/production.js#L32
- `topcoder-react-utils` doesn't have such include paths. If I want to customize webpack config of `connect-app` and add `includePaths` as I have to repeat the whole rule for `.scss` extention https://github.com/appirio-tech/connect-app/blob/dev/config/webpack/production.js#L16-L36 because we use plugin `ExtractCssChunks` and looks like there is no way to override only some parameters of it during webpack config merging. I have to call it again with the full list of parameters.

@birdofpreyru Do you have any idea how to deal with such situations when we need to rewrite only some parameters of plugins?

**Comment by birdofpreyru** [comment]
Yes, in this very case, this entire SCSS rule seems to be the only option. But, if you dublicate it as the rule to apply for `appirio-tech-react-components` only, this makes a perfect sense for me. Sure, we are targeting the reuse of conifgs, but it does not mean that we should not do something like this during transition. As we go forward, we'll just move all components we need into `topcoder-react-ui-kit`, and there they will use the new configs; after that we just drop the `appirio-tech-react-components`.

**Comment by maxceem** [comment]
Yes, makes sense. I've already made some adjustments, if I meet some more difficult situations I'll try to go with dedicated configs approach for `appirio-tech-react-components`.

**Comment by maxceem** [comment]
I have the same issue with `"tc-ui": "git+https://github.com/appirio-tech/tc-ui.git#feature/connectv2"`.
It also doesn't prebuild code, so I have to update it too, to follow `topcoder-react-utils` webpack config.

@vic-topcoder @gondzo What is the plan form `tc-ui`? Is it going to be replaced with `topcoder-react-ui-kit`?

@birdofpreyru looks I cannot make `CSS Modules` ignore third-party packages like `tc-ui` or `appirio-tech-react-components`. 
Even if I ignore these packages in `.scss` rule from `topcoder-react-utils` and make another `.scss` rule only for these packages without `CSS Modules`. CSS modules are still being applied to the styles from these packages. 
I think it's due to the way styles are imported from these third-party modules. We import them in our `.scss` files using `@import` directive rather than importing them in `jsx` files by `import`.

Do you have any ideas, maybe I'm missing something?

**Comment by maxceem** [comment]
Done.

PR #1656 to `connect-app` has to be applied together with [PR 283](https://github.com/appirio-tech/react-components/pull/283) to `react-components` and [PR 22](https://github.com/appirio-tech/tc-ui/pull/22) to `tc-ui`.

I've updated `react-components` and `tc-ui` with the next two things:
- fixed imports to be compatible with `topcoder-react-utils`
- added :global to all scss files to be compatible with CSS Modules from `topcoder-react-utils`

**Comment by maxceem** [comment]
## Note
I haven't updated `eslint` configuration to the one provided by `topcoder-react-utils` because they are based on different presets:
- `topcoder-react-utils` uses `airbnb`
- `connect-app` uses `eslint:recommended`

**Comment by vic-topcoder** [comment]
Does this mean we can start refactoring the button components everywhere in Connect? The other component that would be great to move is Tooltip - after the work we did on Connect, maybe fix the styling and make it available for all apps. This thread is very exciting for me, as we're moving forward unified modern front-end environment... WOOOT! Great job, guys! Mолодцы какие

**Comment by maxceem** [comment]
> Does this mean we can start refactoring the button components everywhere in Connect?

I believe we can. Though I think #1629 has to be finished first. It feels more comfortable to make these tasks one by one as they all introduce a lot of changes across the files.

Btw @vic-topcoder what is the plan for `tc-ui` package? Is it going to be replaced with `topcoder-react-ui-kit`?

**Comment by vic-topcoder** [comment]
> what is the plan for tc-ui package?

Ideally we will break it into 2 - app copy of the styles, and components will be stand-alone (non-dependent on tc-ui)

**Comment by birdofpreyru** [comment]
@maxceem @vic-topcoder 

It looks to me that I got to some milestone :)

From `topcoder-react-utils@0.1.4` you can already adopt all configs (Babel, ESLint, Jest, Stylelint, Webpack), `<Link>`, `<NavLink>` and `<ScalableRect>` helper elements.

From `topcoder-react-ui-kit@0.1.2` you can adopt global styles (actually, just the standard CSS reset + Roboto font declarations) and mixins (typography, media, colors); and themed `<Avatar>`, buttons, and tags.

I've already updated **develop** branch of Community App to use all this stuff from there, and I don't see any issues, at the first glance; though, don't hesitate to ping me if you encounter any problems, or have any questions.

Regarding the next step, I wanna spawn into `topcoder-react-utils` the setup of Express server and the stuff related to the server-side rendering.

**Comment by gondzo** [comment]
continuing the discussion at #1690
